### ISSUE TITLE
Memory cached call and job preemption

### ISSUE THREAD

**Issue by NicolasHug** [joblib/joblib#1030]
Hi all,

I'm posting to test the water about something that I'd be interested to work on. This involves interactions between `Parallel` and `Memory`.

Say I have 10 jobs running in parallel (with joblib). Some of them are calling the same cached function `f` with `Memory`. If 4 jobs are spawned and the cache doesn't exist yet, then `f` is going to be called potentially 4 times. That's 3 unnecessary calls. In addition, some of my jobs *don't* need the output of `f`, so ideally I could just run them while `f` is being computed (there's no point in running the other jobs anyway since they need the output of `f`). Basically what I would need is:

1. a mechanism for a job to find out whether some other job is already doing the same (cached) computation, to avoid redundancy
2. when that happens, let `Parallel` know somehow, and preempt that job to let other jobs be run instead.

My underlying motivation is to avoid unnecessary recomputations when grid searching over scikit-learn pipelines. I think that if we can do 1. and 2., then we can basically do what [dask-ml](https://ml.dask.org/hyper-parameter-search.html#avoid-repeated-work) does (in a different way, but with the same outcome).

I'm not familiar enough with joblib to fully appreciate the technical challenges this might involve, but if that's something that you would consider, I'd be happy to give it a try.

**Comment by pierreglaser** [comment]
Hi, thanks for taking the time to write this up. Could briefly precise what would `f` be in the `GridSearch` so I can get a better grasp of your use case?

**Comment by NicolasHug** [comment]
`f` is [fit_transform_one_cached](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/pipeline.py#L259)

It correspond to the fitting/transform of one pipeline step with a given set of hyper parameters. Or equivalently, it correspond to one node in the DAGs from the dask-ml link avove.

**Comment by pierreglaser** [comment]
Out of curiosity, have you tried to run your example `dask` backend of joblib? I'm still not very familiar with the `dask/joblib` integration but I'm planning to spend most of the next month or so on making it better.

**Comment by NicolasHug** [comment]
I haven't played around with the dask backend yet, but my understanding is that it would be too high-level to do what I want to do. As far as I understand the dask backend would allow to e.g. distribute the jobs on different machines, but what dask would see would be the whole pipelines/jobs, not the individual steps of the pipelines.

**Comment by NicolasHug** [comment]
Hi @ogrisel @GaelVaroquaux @mrocklin ,

@pierreglaser brought to my attention this "Dask/Joblib Integration Roadmap" document, where it seems that similar ideas have been discussed. Would you have any insight to share?

The caching of pipeline steps during a GridSearch is something that @amueller and I would like to work on, so we'd be very interested in joining the discussion!

**Comment by mrocklin** [comment]
> My underlying motivation is to avoid unnecessary recomputations when grid searching over scikit-learn pipelines. I think that if we can do 1. and 2., then we can basically do what dask-ml does (in a different way, but with the same outcome).

I'm obviously biased here, but I'm curious, why not use Dask here?

**Comment by tomMoral** [comment]
Implementing a way to know that specific computation is being run in a robust way can be quite tricky and increase a lot the complexity of the code:

* The classical approach would be to implement a scheduler (something that `dask` does) which optimize which worker runs which job and can thus avoid redundancy. This also include knowing the dependency graph of the function and running functions that do not depends on each other in parallel. This also include pausing the tasks that are waiting on another result and running other task instead. This seems out of scope for a project like `joblib` where the focus is on embarrassingly parallel computations and `dask` is already a good option to do that.

* Another way would be to implement a mechanism in `Memory` that returns a `_CachedFuture` when another worker is already running a job. The `result` method of this future could be blocking while the computation have not complete. Now the tricky questions:
  - how you detect a job is already running.
  - if the job fails, how do you ensure the blocking `result` gets released.
  - how do we do it without breaking the API that the return of the cache function is the result of the function itself.
  - how do we detect `Memory` is being used in a robust way.

  I don't see an answer for the last one. While it is easy to do when the called function is the one in cache, this does not handle the nested cached function and I don't see a good way to detect this. And this happens often. Maybe this is not a central question and we can say this only works when the cache function is being called directly. But I think in the `GridSearch` case, what we would like to be cached and reused could be preprocessing steps in the pipeline that are not run at top level no?

* Finally, another way would be to manage to "pause" the computations when a worker needs to run a cached function that is already being run. The worker would then be able take run another job. This would allow for decentralized job management. But this is quite a change in paradigm and does not fit in the `Pool/Executor` implementations. This thus needs a lot of work to allow the worker to resume the computation in a robust way. Also, making sure we do not create deadlocks and freeze is not so easy.

In the current state of affairs, I would say the good solution is using `dask` as it already provides a scheduler which is able to do this in a centralized way. These are some elements of my reflexion on the subject, I am more than happy to discuss about them!

**Comment by amueller** [comment]
@mrocklin if there's a relatively non-intrusive way to use dask, than that would be great. Or do you think we should just implement the same approach as daskml?

Also, it's an open question whether we want to add a dask dependency. I'm not against it's something to consider.

For option 2:
- Detecting if a job is running could be done via file system, right? Is that a bad idea?
- Failure I have no idea. That seems hard?
- I'm not sure I understand the third and fourth point.

The difference between 2 and 3is how the worker pool is managed, right? But all the problems of 
2 also apply to 3, right?

It sounded to me like option 2 was what was discussed in the joblib / dask integration roadmap, but if I misread that, then I'd be happy to be educated :)
I just wanted to check in what's on the roadmap there so we can plan accordingly.

**Comment by mrocklin** [comment]
> if there's a relatively non-intrusive way to use dask, than that would be great

Ah, are you saying that this is for sklearn itself?  I was assuming that this was for something external. 
 
I'm not qualified to answer the dependency question.  To argue slightly for it though, I'll suggest that you could always vendor in the Dask local scheduler.  It's probably as-or-less complex than joblib.  If you can get everything you want with caching then awesome.  If you decide that you need proper dependency handling then I suspect that you would end up coming to roughly a Dask-like implementation.

**Comment by amueller** [comment]
@mrocklin yes, I think this particular feature of dask-ml is so awesome that I want it in sklearn ;)

**Comment by mrocklin** [comment]
I want it in sklearn too, so that you all think about it instead of Dask folk.  (you all are better at this than we are)
